src/main/java/org/opensearch/ad/rest/handler/ModelValidationActionHandler.java

```diff
@@ -0,0 +1,718 @@
+            if (anomalyDetector.getCategoryField().size() == 1) {
+                Terms entities = aggs.get(AGG_NAME_TOP);
+                Object key = entities
+                    .getBuckets()
+                    .stream()
+                    .max(Comparator.comparingInt(entry -> (int) entry.getDocCount()))
+                    .map(MultiBucketsAggregation.Bucket::getKeyAsString)
+                    .orElse(null);
+                topKeys.put(anomalyDetector.getCategoryField().get(0), key);
+            } else {
+                CompositeAggregation compositeAgg = aggs.get(AGG_NAME_TOP);
+                topKeys
+                    .putAll(
+                        compositeAgg
+                            .getBuckets()
+                            .stream()
+                            .flatMap(bucket -> bucket.getKey().entrySet().stream()) // this would create a flattened stream of map entries
+                            .collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue()))
+                    );
+            }
+            for (Map.Entry<String, Object> entry : topKeys.entrySet()) {
+                if (entry.getValue() == null) {
+                    topEntityListener.onResponse(Collections.emptyMap());
+                    return;
+                }
+            }
+            topEntityListener.onResponse(topKeys);
+        }, topEntityListener::onFailure));
+    }
+
+    private void getLatestDateForValidation(Map<String, Object> topEntity) {
+        ActionListener<Optional<Long>> latestTimeListener = ActionListener
+            .wrap(latest -> getSampleRangesForValidationChecks(latest, anomalyDetector, listener, topEntity), exception -> {
+                listener.onFailure(exception);
+                logger.error("Failed to create search request for last data point", exception);
+            });
+        searchFeatureDao.getLatestDataTime(anomalyDetector, latestTimeListener);
+    }
+
+    private void getSampleRangesForValidationChecks(
+        Optional<Long> latestTime,
+        AnomalyDetector detector,
+        ActionListener<ValidateAnomalyDetectorResponse> listener,
+        Map<String, Object> topEntity
+    ) {
+        if (!latestTime.isPresent()) {
+            listener
+                .onFailure(
+                    new ADValidationException(
+                        CommonErrorMessages.NOT_ENOUGH_HISTORICAL_DATA,
+                        DetectorValidationIssueType.GENERAL_DATA,
+                        ValidationAspect.MODEL
+                    )
+                );
+            return;
+        } else if (latestTime.get() <= 0) {
+            listener
+                .onFailure(
+                    new ADValidationException(
+                        CommonErrorMessages.NOT_ENOUGH_HISTORICAL_DATA,
+                        DetectorValidationIssueType.GENERAL_DATA,
+                        ValidationAspect.MODEL
+                    )
+                );
+            return;
+        }
```

Comment:
Seems these 2 can be consolidated in one block. Also, when would the time be <= 0?

Reply:
Consildating these into 1. There are cases when last timestamp is negative when date field is in index mapping but its not ever filled (these happens on when timefield is set to `customer_birth_date` for the opensearch dashboard sample data). It would return a last timestamp of `-9223372036854775808`